File: node/src/service/provider.ts
import { GrpcProviderAdaptee } from '../lib/grpc/adaptee';
import { Any } from 'google-protobuf/google/protobuf/any_pb';
import { SERVICE_TYPE, SERVICE_TYPEMap, ZDResponse } from '../lib/proto/wrappers_pb';
import { ZDServiceRequest, ZDServiceRequestResult, ZDService } from '../lib/proto/manager_pb'
import { BlockFilter, ProtocolTag } from '../lib/proto/networkif_pb';
import logger from '../lib/utils/logger';

export default class Provider extends GrpcProviderAdaptee {

    public getSubscriberId(): string {
        return 'provider:9999'
    }

    public getGrpcHost(): string {
        return 'localhost:5000';
    }

    public getSubscriberServiceTypes(): SERVICE_TYPEMap[keyof SERVICE_TYPEMap][] {
        return [
            SERVICE_TYPE.SERVICE_SET_NETWORK
        ];
    }

    public getCurrentServiceStatus(): ZDService[] {
        return [
            this.newZDService(SERVICE_TYPE.SERVICE_SET_NETWORK, ZDService.SERVICE_STATUS.ENABLED)
        ]
    }

    public onSubscriberServiceRequest(req: ZDServiceRequest): ZDServiceRequestResult | null {
        let result = null;
        const session: string = req.getSession();
        logger.info('receive service req :', req.getType(), req.getSession());
        if (req && session) {
            const type: SERVICE_TYPEMap[keyof SERVICE_TYPEMap] = req.getType();
            switch (type) {
                case SERVICE_TYPE.SERVICE_SYS_CANCEL:
                    logger.info(`req:${req.getSession()} timeout.`);
                    break;
                case SERVICE_TYPE.SERVICE_SET_NETWORK:
                    result = this.onHandleExampleServiceReq(req);
                    break;
            }
        }
        return result;
    }

    private onHandleExampleServiceReq(req: ZDServiceRequest): ZDServiceRequestResult | null {
        const data: Any | undefined = req.getData();
        if (data) {
            const args = data.unpack(BlockFilter.deserializeBinary, 'zdautomotive.protobuf.BlockFilter');
            if (args) {
                logger.info('unpack req args data:',
                    args.getPhyid(),
                    args.getVlanid(),
                    args.getSrcip(),
                    args.getDstip(),
                    args.getSrcport(),
                    args.getDstport());
                const result: ZDServiceRequestResult = new ZDServiceRequestResult();
                const response: ZDResponse = new ZDResponse();
                response.setCode(ZDResponse.ERROR_CODE.OK);
                response.setMessage('send req success.');
                const anyData: Any = new Any();
                const parameter: ProtocolTag = new ProtocolTag();
                parameter.setVlanid(456);
                parameter.setSrcport(456);
                parameter.setDstport(456);
                parameter.setTag('456');
                anyData.pack(parameter.serializeBinary(), 'zdautomotive.protobuf.ProtocolTag');
                response.setData(anyData);
                result.setResponse(response);
                result.setRequest(req);
                return result;
            }
        }
        return null;
    }

    private newZDService(type: SERVICE_TYPEMap[keyof SERVICE_TYPEMap], status: ZDService.SERVICE_STATUSMap[keyof ZDService.SERVICE_STATUSMap] = ZDService.SERVICE_STATUS.ENABLED): ZDService {
        const service: ZDService = new ZDService();
        service.setType(type);
        service.setStatus(status);
        return service;
    }

}

File: cpp/src/service/provider.cpp
#include <iostream>
#include "provider.h"

Provider ::Provider()
{
    // cout << "Provider" << endl;
    p_service_type = new SERVICE_TYPE[SUB_SERVICE_NUM];
    p_service_type[0] = SERVICE_TYPE::SERVICE_SET_NETWORK;
    p_service_status = new ZDService[SUB_SERVICE_NUM];
    ZDService service;
    service.set_type(SERVICE_TYPE::SERVICE_SET_NETWORK);
    service.set_status(ZDService_SERVICE_STATUS::ZDService_SERVICE_STATUS_ENABLED);
    p_service_status[0] = service;
}

Provider ::~Provider()
{
    // cout << "~Provider" << endl;
    delete[] p_service_type;
    delete[] p_service_status;
}

string Provider ::getSubscriberId() const
{
    return "provider:8888";
}

string Provider ::getGrpcHost() const
{
    return "localhost:5000";
}

ZDServiceSubscriber_SUBSCRIBER_TYPE Provider::getSubscriberType() const
{
    return ZDServiceSubscriber_SUBSCRIBER_TYPE::ZDServiceSubscriber_SUBSCRIBER_TYPE_PROVIDER;
}

SERVICE_TYPE *Provider ::getSubscriberServiceTypes() const
{
    return p_service_type;
}

ZDService *Provider ::getCurrentServiceStatus() const
{
    return p_service_status;
}

int Provider::getSubscriberServiceTypesNum() const
{
    return SUB_SERVICE_NUM;
}

ZDServiceRequestResult Provider::onSubscriberServiceRequest(ZDServiceRequest &req) const
{
    ZDServiceRequestResult result;
    SERVICE_TYPE sub_type = req.type();
    cout << "receive service req :" << sub_type << endl;
    switch (sub_type)
    {
    case SERVICE_TYPE::SERVICE_SET_NETWORK:
        onHandleSetNetworkReq(req, &result);
        break;
    }
    return result;
}

void Provider::onHandleSetNetworkReq(ZDServiceRequest &req, ZDServiceRequestResult *result) const
{
    cout << "unpack req any data ..." << endl;
    Any *p = req.mutable_data();
    Any reqAny = *p;
    BlockFilter filter;
    if (reqAny.UnpackTo(&filter))
    {
        cout << "phyid:" << filter.phyid() << endl;
        cout << "vlanid:" << filter.vlanid() << endl;
        cout << "srcip:" << filter.srcip() << endl;
        cout << "dstIp:" << filter.dstip() << endl;
        cout << "srcport:" << filter.srcport() << endl;
        cout << "dstport:" << filter.dstport() << endl;
    }
    ZDResponse res;
    res.set_code(ZDResponse_ERROR_CODE::ZDResponse_ERROR_CODE_OK);
    res.set_message("send req successfully.");
    ProtocolTag tag;
    tag.set_vlanid(456);
    tag.set_srcport(456);
    tag.set_dstport(466);
    tag.set_tag("465");
    Any resAny;
    resAny.PackFrom(tag);
    res.mutable_data()->MergeFrom(resAny);
    result->mutable_request()->MergeFrom(req);
    result->mutable_response()->MergeFrom(res);
}